File: NapakalakiRuby/lib/CardDealer.rb
#encoding: UTF-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require "singleton"
require_relative "Prize.rb"
require_relative "TreasureKind.rb"
require_relative "BadConsequence.rb"
require_relative "Monster.rb"
require_relative "cultist.rb"
require_relative "specific_bad_consecuence.rb"
require_relative "death_bad_consequence.rb"
require_relative "numeric_bad_consequence.rb"

module NapakalakiGame

	class CardDealer
		include Singleton
    
    attr_accessor :used_monster, :unused_monster, :used_treasures, :unused_treasures
    
    def initialize
      @unused_treasures = Array.new
      @used_treasures = Array.new
      @used_monster = Array.new
      @unused_monster = Array.new
      @unused_cultist = Array.new
      
    end
    
	def initTreasureCardDeck      
      @unused_treasures << Treasure.new("Si mi amo", 4, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("Botas de investigacion", 3, [TreasureKind::SHOES])
      @unused_treasures << Treasure.new("Capucha de Cthulhu", 3, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("A prueba de babas", 2, [TreasureKind::ARMOR])
      @unused_treasures << Treasure.new("Botas de lluvia acida", 1, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Casco minero", 2, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("Ametralladora Thompson", 4, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Camiseta de la UGR", 1, [TreasureKind::ARMOR])
      @unused_treasures << Treasure.new("Clavo del rail ferroviario", 3, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Cuchillo de sushi arcano", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Fez alopodo", 3, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("Hacha prehistorica", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("El aparato del Pr. Tesla", 4, [TreasureKind::ARMOR])
      @unused_treasures << Treasure.new("Gaita", 4, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Insecticida", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Escopeta de 3 cañones", 4, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Garabato mistico", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("La rebeca metalica", 2, [TreasureKind::ARMOR])
      @unused_treasures << Treasure.new("Lanzallamas", 4, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Necrocomicon", 1, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Necronomicon", 5, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Linterna a dos manos", 3, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Necrognomicon", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Necrotelecom", 2, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("Mazo de los antiguos", 3, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Necroplayboycon", 3, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Porra preternatural", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Shogulador", 1, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Varita de atizamiento", 3, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Tentaculo de pega", 2, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("Zapato deja-amigos", 1, [TreasureKind::SHOES])
      
	end

	def initMonsterCardDeck
     
        prize = Prize.new(2,1)
        badConsequence = SpecificBadConsecuence.new("Pierdes tu armadura visible y otra oculta",  0,[TreasureKind::ARMOR],[TreasureKind::ARMOR])
        unused_monster<< Monster.new("Byakhees de bonanza", 8, prize, badConsequence, 0)

        prize = Prize.new(1,1)
        badConsequence = SpecificBadConsecuence.new("Embobados con el lindo primigenio te descartas de tu casco visible", 0, [TreasureKind::ARMOR], [])
        unused_monster<< Monster.new("Chibithulhu", 2, prize, badConsequence, 0)

        prize = Prize.new(1,1)
        badConsequence = SpecificBadConsecuence.new("El primordial bostezo contagioso. Pierdes el calzado visible", 0, [TreasureKind::SHOES], [])
        unused_monster<< Monster.new("El sopor de Dunwich", 2, prize, badConsequence, 0)

        prize = Prize.new(4,1)
        badConsequence = SpecificBadConsecuence.new("Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y 1 mano oculta", 0, [TreasureKind::ONEHAND], [TreasureKind::ONEHAND])
        unused_monster<< Monster.new("Ángeles de la noche ibicenca", 14, prize, badConsequence, 0)

        prize = Prize.new(3,1)
        badConsequence = NumericBadConsequence.new("Pierdes todos los tesoros visibles", 0,5,0)
        unused_monster<< Monster.new("El gorrón en el umbral", 10, prize, badConsequence, 0)

        prize = Prize.new(2,1)
        badConsequence = SpecificBadConsecuence.new("Pierdes la armadura visible", 0, [TreasureKind::ARMOR], [])
        unused_monster<< Monster.new("H.P. Munchcraft", 6 , prize, badConsequence, 0)

        prize = Prize.new(1,1)
        badConsequence = SpecificBadConsecuence.new("Sientes bichos bajo la ropa. Descarta la armadura visible", 0, [TreasureKind::ARMOR], [])
        unused_monster<< Monster.new("Bichgooth",2,prize,badConsequence, 0)

        price = Prize.new(4,2) 
        badConsequence = NumericBadConsequence.new("Pierdes 5 niveles y 3 tesoros visibles",5 , 3, 0)
        unused_monster << Monster.new("El rey de rosa",13,prize,badConsequence, 0)

        prize = Prize.new(1,1)
        badConsequence = NumericBadConsequence.new("Toses los pulmones y pierdes 2 niveles",2, 0, 0)
        unused_monster<< Monster.new("La que redacta las tinieblas",2, prize, badConsequence, 0)

        prize = Prize.new(1,1)
        badConsequence = SpecificBadConsecuence.new("Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos", 3, [TreasureKind::BOTHHANDS], [])
        unused_monster<< Monster.new("Bicéfalo", 20, prize, badConsequence, 0)

        #Los Hondos
        prize = Prize.new(2, 1)
        badConsequence = DeathBadConsequence.new("Estos monstruos resultan bastante superficiales y te aburren mortalmente. Estas muerto", )
        unused_monster<< Monster.new("Los hondos", 8, prize, badConsequence, 0)

        #Semillas Cthulhu
        prize = Prize.new(2, 1)
        badConsequence = NumericBadConsequence.new("Pierdes 2 niveles y dos tesoros ocultos", 2, 2, 0)
        unused_monster<< Monster.new("Semillas Cthulhu", 4, prize, badConsequence, 0)

        #Dameargo
        prize = Prize.new(2, 1)
        badConsequence = SpecificBadConsecuence.new("Te intentas escaquear. Pierdes una mano visible", 0, [TreasureKind::ONEHAND], [])
        unused_monster<< Monster.new("Dameargo", 1, prize, badConsequence, 0)

        #Pollipolipo Volante
        prize = Prize.new(1, 1)
        badConsequence = NumericBadConsequence.new("Da mucho asquito. Pierdes 3 niveles", 3, 0, 0)
        unused_monster<< Monster.new("Pollipolipo Volante", 3, prize, badConsequence, 0)

        #Yskhtihyssg-Goth
        prize = Prize.new(3, 1)
        badConsequence = DeathBadConsequence.new("No le hace gracia que pronuncien mal su nombre. Estas muerto")
        unused_monster<< Monster.new("Yskhtihyssg-Goth", 12, prize, badConsequence, 0)

        #Familia Feliz
        prize = Prize.new(4, 1)
        badConsequence = DeathBadConsequence.new("La familia te atrapa. Estas muerto")
        unused_monster<< Monster.new("Familia Feliz", 1, prize, badConsequence, 0)

        #Roboggoth
        prize = Prize.new(2, 1)
        badConsequence = NumericBadConsequence.new("La quinta directiva primaria te obliga a perder 2 niveles y un tesoro 2 manos visibles", 2, 2, 0)
        unused_monster<< Monster.new("Roboggoth", 8, prize, badConsequence, 0)

        #El Espia
        prize = Prize.new(1, 1)
        badConsequence = SpecificBadConsecuence.new("Te asusta en la noche. Pierdes un casco visible", 0, [TreasureKind::ONEHAND], [])
        unused_monster<< Monster.new("El Espia", 5, prize, badConsequence, 0)

        #El Lenguas
        prize = Prize.new(1, 1)
        badConsequence = NumericBadConsequence.new("Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles", 2, 5, 0)
        unused_monster<< Monster.new("El Lenguas", 20, prize, badConsequence, 0)
        
        #**********************************************************Monstruos con Sectarios**********************************************************#

        prize = Prize.new(3, 1)
        badConsequence = NumericBadConsequence.new("Pierdes una mano visible", 0, 2, 0)
        unused_monster << Monster.new("El mal indecible impronunciable", 10, prize, badConsequence, -2)

        prize = Prize.new(2, 1)
        badConsequence = SpecificBadConsecuence.new("Pierdes tus tesoros visibles. Jajaja", 0, nil, 0)
        unused_monster << Monster.new("Testigos Oculares", 6, prize, badConsequence, +2)

        prize = Prize.new(2, 5)
        badConsequence = DeathBadConsequence.new("Hoy no es tu dia de suerte. Mueres")
        unused_monster << Monster.new("El gran Cthulhu", 20, prize, badConsequence, +4)

        prize = Prize.new(2, 1)
        badConsequence = NumericBadConsequence.new("Tu gobierno te recorta 2 niveles", 2, 0, 0)
        unused_monster << Monster.new("Serpiente Politico", 8, prize, badConsequence, -2)

        prize = Prize.new(1, 1)
        badConsequence = SpecificBadConsecuence.new("Pierdes tu casco y tu armadura visible. Pierdes tus manos ocultas", 0, [TreasureKind::HELMET, TreasureKind::ARMOR], nil)
        unused_monster << Monster.new("Felpuggoth", 2, prize, badConsequence, +5)

        prize = Prize.new(4, 2)
        badConsequence = NumericBadConsequence.new("Pierdes dos niveles", 2, 0, 0)
        unused_monster << Monster.new("Shoggoth", 16, prize, badConsequence, -4)

        prize = Prize.new(1, 1)
        badConsequence = NumericBadConsequence.new("Pintalabios negro. Pierdes 2 niveles", 2, 0, 0)
        unused_monster << Monster.new("Lolitagooth", 2, prize, badConsequence, +3)
	end

	def shuffleTreasures
      @unused_treasures = @unused_treasures.shuffle

	end
  
  def initCultistCardDeck
    @unused_cultist << Cultist.new("Sectario: +1 por cada sectario en juego. No puedes dejar de ser sectario.", 1)
    @unused_cultist << Cultist.new("Sectario: +2 por cada sectario en juego. No puedes dejar de ser sectario.", 2)
    @unused_cultist << Cultist.new("Sectario: +1 por cada sectario en juego. No puedes dejar de ser sectario.", 1)
    @unused_cultist << Cultist.new("Sectario: +2 por cada sectario en juego. No puedes dejar de ser sectario.", 2)
    @unused_cultist << Cultist.new("Sectario: +1 por cada sectario en juego. No puedes dejar de ser sectario.", 1)
    @unused_cultist << Cultist.new("Sectario: +1 por cada sectario en juego. No puedes dejar de ser sectario.", 1)
  end

	def shuffleMonster
      @unused_monster = @unused_monster.shuffle

	end
  
  def shuffleCultist
    @unused_cultist = @unused_cultist.shuffle
  end

	def getInstance
		@instance
	end

	def nextTreasure
      if @unused_treasures.nil?
        @used_treasures.each do |t|
          @unused_treasures<<t
        end

	      shuffleTreasures
	      @used_treasure.clear
      end
      t = @unused_treasures.at(0)
      @used_treasures<<t
      @unused_treasures.delete(t)
      return t

	end

	def nextMonster
      if @unused_monster.nil?
        @used_monster.each do |m|
         	@unused_monster<<m
        end
        	shuffleMonster
        	@used_monster.clear
      end

	  m = @unused_monster.at(0)
	  @unused_monster.delete(m)
	  @used_monster<<m

      return m

	end

	def giveTreasureBack(treasure)
		if !used_treasures.include?
      		@used_treasures << treasure
      	end
	end

	def giveMonsterBack(monster)
		if !used_monster.include?
      		@used_monster << monster
      	end
	end

	def initCards
	    initTreasureCardDeck
	    shuffleTreasures
	    
	    initMonsterCardDeck
	    shuffleMonster
      
    initCultistCardDeck
    shuffleCultist
	end
  end
end

File: Napakalaki/src/NapakalakiGame/CardDealer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Arrays;


/**
 *
 * @author david
 */
public class CardDealer {
    	private static CardDealer instance = null;
        private ArrayList<Treasure> unusedTreasures= new ArrayList();
        private ArrayList<Monster> usedMonster= new ArrayList();
        private ArrayList<Treasure> usedTreasures= new ArrayList();
        private ArrayList<Monster> unusedMonster= new ArrayList();
        private ArrayList<Cultist> cultistDeck= new ArrayList();
        
        

	private CardDealer(){

	}

	private void initTreasureCardDeck(){
           unusedTreasures = new ArrayList();
           
           unusedTreasures.add(new Treasure("Si mi amo", 4, TreasureKind.HELMET));
           unusedTreasures.add(new Treasure("Botas de investigacion", 3, TreasureKind.SHOES));
           unusedTreasures.add(new Treasure("Capucha de Cthulhu", 3, TreasureKind.HELMET));
           unusedTreasures.add(new Treasure("A prueba de babas", 2, TreasureKind.ARMOR));
           unusedTreasures.add(new Treasure("Botas de lluvia acida", 1, TreasureKind.BOTHHANDS));
           unusedTreasures.add(new Treasure("Casco minero", 2, TreasureKind.HELMET));
           unusedTreasures.add(new Treasure("Ametralladora Thompson", 4, TreasureKind.BOTHHANDS));
           unusedTreasures.add(new Treasure("Camiseta de la UGR", 1, TreasureKind.ARMOR));
           unusedTreasures.add(new Treasure("Clavo de rail ferroviario", 3, TreasureKind.ONEHAND));
           unusedTreasures.add(new Treasure("Cuchillo de sushi arcano", 2, TreasureKind.ONEHAND));
           unusedTreasures.add(new Treasure("fez alopodo", 3, TreasureKind.HELMET));
           unusedTreasures.add(new Treasure("Hacha prehistorica", 2, TreasureKind.ONEHAND));
           unusedTreasures.add(new Treasure("El aparato del Pr. Tesla", 4, TreasureKind.ARMOR));
           unusedTreasures.add(new Treasure("Gaita", 4, TreasureKind.BOTHHANDS));
           unusedTreasures.add(new Treasure("Insecticida", 2, TreasureKind.ONEHAND));
           unusedTreasures.add(new Treasure("Escopeta de 3 cañones", 4, TreasureKind.BOTHHANDS));
           unusedTreasures.add(new Treasure("Garabato mistico", 2, TreasureKind.ONEHAND));
           unusedTreasures.add(new Treasure("La rebeca metalica", 2, TreasureKind.ARMOR));
           unusedTreasures.add(new Treasure("Lanzallamas", 4, TreasureKind.BOTHHANDS));
           unusedTreasures.add(new Treasure("Necrocomicon", 1, TreasureKind.ONEHAND));
           unusedTreasures.add(new Treasure("Necronomicon", 5, TreasureKind.BOTHHANDS));
           unusedTreasures.add(new Treasure("Linterna a dos manos", 3, TreasureKind.BOTHHANDS));
           unusedTreasures.add(new Treasure("Necrognomicon", 2, TreasureKind.ONEHAND));
           unusedTreasures.add(new Treasure("Necrotelecom", 2, TreasureKind.HELMET));
           unusedTreasures.add(new Treasure("Mazo de los antiguos", 3, TreasureKind.ONEHAND));
           unusedTreasures.add(new Treasure("Necroplayboycon", 3, TreasureKind.ONEHAND));
           unusedTreasures.add(new Treasure("Porra preternatural", 2, TreasureKind.ONEHAND));
           unusedTreasures.add(new Treasure("Shogulador", 1, TreasureKind.BOTHHANDS));
           unusedTreasures.add(new Treasure("Varita de atizamiento", 3, TreasureKind.ONEHAND));
           unusedTreasures.add(new Treasure("Tentaculo de pega", 2, TreasureKind.HELMET));
           unusedTreasures.add(new Treasure("Zapato deja-amigos", 1, TreasureKind.SHOES));
	}

	private void initMonsterCardDeck(){
                    //bonanza
        BadConsequence badConsequence = new SpecificBadConsequence("Pierdes tu armadura visible y otra oculta", 0, new ArrayList(Arrays.asList(TreasureKind.ARMOR)), new ArrayList(Arrays.asList(TreasureKind.ARMOR)));
        Prize prize = new Prize(2, 1);
        unusedMonster.add(new Monster("Byakhees de bonanza", 8, prize, badConsequence));
        
        //Chibichulo
        badConsequence = new SpecificBadConsequence("Embobados con ellindo primigenio te descartas tu casco visible", 0, new ArrayList(Arrays.asList(TreasureKind.HELMET)), new ArrayList());
        prize = new Prize(1, 1);        
        unusedMonster.add(new Monster("Chibithulhu", 2, prize, badConsequence));
        
        //Dunwich
        badConsequence = new SpecificBadConsequence("El primordial bostezo contagioso. Pierdes el calzado visible", 0, new ArrayList(Arrays.asList(TreasureKind.SHOES)), new ArrayList());
        prize = new Prize(1, 1);        
        unusedMonster.add(new Monster("El sopor de Dunwitch", 2, prize, badConsequence));
        
        //Angeles de la noche ibicenca
        badConsequence = new SpecificBadConsequence("Te atrapan para llevarte a una fiesta y te dejan caer en mitad del vuelo. Descarta una mano visible y una mano oculta", 0, new ArrayList(Arrays.asList(TreasureKind.ONEHAND)), new ArrayList());
        prize = new Prize(3, 1);        
        unusedMonster.add(new Monster("Ángeles de la noche ibicenca", 14, prize, badConsequence));
        
        //El gorron en el umbral      Pendiente de revision
        badConsequence = new NumericBadConsequence("Pierdes todos tus tesoros visibles", 0, 5, 0);
        prize = new Prize(3, 1);
        unusedMonster.add(new Monster("El gorrón en el umbral", 10, prize, badConsequence));
        
        //munchcraft
        badConsequence = new SpecificBadConsequence("Pierdes la armadura visible", 0, new ArrayList(Arrays.asList(TreasureKind.ARMOR)), new ArrayList());
        prize = new Prize(2, 1);
        unusedMonster.add(new Monster("H.P. Munchcraft", 6, prize, badConsequence));
        
        //bichgooth
        badConsequence = new SpecificBadConsequence("Sientes bichos bajo la ropa. Descarta la armadura visible", 0, new ArrayList(Arrays.asList(TreasureKind.ARMOR)), new ArrayList());
        prize = new Prize(1, 1);
        unusedMonster.add(new Monster("Bichgooth", 2, prize, badConsequence));
        
        //Rey de rosa
        badConsequence = new NumericBadConsequence("Pierdes 5 niveles y 3 tesoros visibles", 5, 3, 0);
        prize = new Prize(4, 2);
        unusedMonster.add(new Monster("El rey de rosa", 13, prize, badConsequence));
        
        //La que redacta las tinieblas
        badConsequence = new NumericBadConsequence("Toses los pulmones y pierdes 2 niveles", 2, 0, 0);
        prize = new Prize(1, 1);
        unusedMonster.add(new Monster("La que redacta en las tinieblas",2 , prize, badConsequence));
        
        //Bicefalo
        badConsequence = new SpecificBadConsequence("Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos", 3, new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS)), new ArrayList());
        prize = new Prize(1, 1);
        unusedMonster.add(new Monster("Bicéfalo", 20, prize, badConsequence));
        
        //Los Hondos
        badConsequence = new DeathBadConsequence("Estos unusedMonster resultan bastante superficiales y te aburren mortalmente. Estas muerto", true);
        prize = new Prize(2, 1);
        unusedMonster.add(new Monster("Los hondos", 8, prize, badConsequence));
        
        //Semillas Cthulhu
        badConsequence = new NumericBadConsequence("Pierdes 2 niveles y 2 tesoros ocultos", 2, 0, 2);
        prize = new Prize(2, 1);
        unusedMonster.add(new Monster("Semillas Cthulhu", 4, prize, badConsequence));
        
        //Dameargo
        badConsequence = new SpecificBadConsequence("Te intentas escaquear. Pierdes una mano invisible", 0, new ArrayList(Arrays.asList(TreasureKind.ONEHAND)), new ArrayList());
        prize = new Prize(2, 1);
        unusedMonster.add(new Monster("Dameargo", 1, prize, badConsequence));
        
        //Pollipolipo Volante
        badConsequence = new NumericBadConsequence("Da mucho asquito. Pierdes 3 niveles", 3, 0, 0);
        prize = new Prize(1, 1);
        unusedMonster.add(new Monster("Pollipolipo volante", 3, prize, badConsequence));
        
        //Yskhtihussg-Goth
        badConsequence = new DeathBadConsequence("No le hace gracia que pronuncien mal su nombre. Estas muerto", true);
        prize = new Prize(3, 1);
        unusedMonster.add(new Monster("Yskhtihussg-Goth", 12, prize, badConsequence));
        
        //Familia Feliz
        badConsequence = new DeathBadConsequence("La familia te atrapa. Estas muerto", true);
        prize = new Prize(4, 1);
        unusedMonster.add(new Monster("Familia feliz", 1, prize, badConsequence));
        
        //Roboggoth
        badConsequence = new SpecificBadConsequence("La quinta directiva primaria te obliga a perder 2 niveles y un tesoro 2 manos visible", 2, new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS)), new ArrayList());
        prize = new Prize(2, 1);
        unusedMonster.add(new Monster("Roboggoth", 8, prize, badConsequence));
        
        //El Espia
        badConsequence = new SpecificBadConsequence("Te asusta en la noche. Pierdes un casco visible", 0, new ArrayList(Arrays.asList(TreasureKind.HELMET)), new ArrayList());
        prize = new Prize(1, 1);
        unusedMonster.add(new Monster("El espia", 5, prize, badConsequence));
        
        //El Lenguas
        badConsequence = new NumericBadConsequence("Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles", 2, 5, 0);
        prize = new Prize(1, 1);
        unusedMonster.add(new Monster("El Lenguas", 20, prize, badConsequence));

                /**********************************************************Monstruos con Sectarios**********************************************************/

        badConsequence = new SpecificBadConsequence("Pierdes una mano visible", 0, new ArrayList(Arrays.asList(TreasureKind.ONEHAND)), new ArrayList());
        prize = new Prize(3, 1);
        unusedMonster.add(new Monster("El mal indecible inpronunciable", 10, prize, badConsequence, -2));

        badConsequence = new NumericBadConsequence("Pierdes tus tesoreos visibles. Jajaja", 0, Integer.MAX_VALUE, 0);
        prize = new Prize(2, 1);
        unusedMonster.add(new Monster("Testigos Oculares", 6, prize, badConsequence, +2));

        badConsequence = new DeathBadConsequence("Hoy no es tu dia de suerte. Mueres", true);
        prize = new Prize(2, 5);
        unusedMonster.add(new Monster("El gran Cthulhu", 20, prize, badConsequence, +4));

        badConsequence = new NumericBadConsequence("Tu gobierno te recorta 2 niveles", 2, 0, 0);
        prize = new Prize(2, 1);
        unusedMonster.add(new Monster("Serpiente Politico", 8, prize, badConsequence, -2));

        badConsequence = new SpecificBadConsequence("Pierdes tu casco y tu armadura visible. Pierdes tus manos ocultas", 0, new ArrayList(Arrays.asList(
          TreasureKind.HELMET, TreasureKind.ARMOR)), new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS, TreasureKind. ONEHAND)));
        prize = new Prize(1, 1);
        unusedMonster.add(new Monster("Felpugoth", 2, prize, badConsequence, +5));

        badConsequence = new NumericBadConsequence("Pierdes dos niveles", 2, 0, 0);
        prize = new Prize(4, 2);
        unusedMonster.add(new Monster("Shoggoth", 16, prize, badConsequence, -4));

        badConsequence = new NumericBadConsequence("Pintalabios negro. Pierdes dos niveles", 2, 0, 0);
        prize = new Prize(1, 1);
        unusedMonster.add(new Monster("Lolitagooth", 2, prize, badConsequence, +3));
	}

        
        private void initCultistCardDeck(){

            cultistDeck.add( new Cultist("Sectario: +1 por cada sectario en juego. No puedes dejar de ser sectario", 1));

            cultistDeck.add( new Cultist("Sectario: +2 por cada sectario en juego. No puedes dejar de ser sectario", 2));

            cultistDeck.add( new Cultist("Sectario: +1 por cada sectario en juego. No puedes dejar de ser sectario", 1));

            cultistDeck.add( new Cultist("Sectario: +2 por cada sectario en juego. No puedes dejar de ser sectario", 2));

            cultistDeck.add( new Cultist("Sectario: +1 por cada sectario en juego. No puedes dejar de ser sectario", 1));

            cultistDeck.add( new Cultist("Sectario: +1 por cada sectario en juego. No puedes dejar de ser sectario", 1));

            shuffleCultist();
        }


	private void shuffleTreasures(){
            Collections.shuffle(this.unusedTreasures);
	}

	private void shuffleMonsters(){
            Collections.shuffle(this.unusedMonster);
	}
        
        private void shuffleCultist(){
            Collections.shuffle(this.cultistDeck);
        }

	public static CardDealer getInstance(){
		if(instance == null){
			instance = new CardDealer();
		}
		return instance;
	}

	public Treasure nextTreasure(){
            if(this.unusedTreasures == null){
              for(Treasure t: this.usedTreasures){
                this.unusedTreasures.add(t);
              }

              this.shuffleTreasures();
              this.usedTreasures.clear();
            }
            Treasure t = this.unusedTreasures.get(0);
            this.usedTreasures.add(t);
            this.unusedTreasures.remove(t);

            return t;

        }

        public Monster nextMonster(){
            if(this.unusedMonster == null){
                for(Monster m: this.usedMonster){
                    this.unusedMonster.add(m);
            }

            this.shuffleMonsters();
            this.usedMonster.clear();
            }
            Monster m = this.unusedMonster.get(0);
            this.usedMonster.add(m);
            this.unusedMonster.remove(m);
            return m;
        }
        
        public Cultist nextCultist(){
            Cultist c;
            c = cultistDeck.get(0);
            cultistDeck.remove(0);
            return c;
            
        }

        public void giveTreasureBack(Treasure t){
            this.usedTreasures.add(t);
        }

        public void giveMonsterBack(Monster m){
                this.usedMonster.add(m);
        }

        public void initCards(){            
            this.initTreasureCardDeck();
            this.initMonsterCardDeck();
            this.shuffleMonsters();
            this.shuffleTreasures();
            this.shuffleCultist();

        }

   
}
